hpc0.qmd

```diff
@@ -103,195 +115,513 @@ ssh <username>@arc4.leeds.ac.uk
+
 :::
 
-## Exercise 4 {.smaller}
+
+## sed 1/2
+
+`sed` reads an input stream and produces an output stream. Internally it has a pattern space and a hold buffer. Data is read from the input stream until the next newline character and placed into the pattern space. Most sed commands operate on the data in the pattern space. Data can be exchanged between the pattern space and the hold buffer. Once sed has executed all commands, it outputs the pattern space and adds a newline characer at the end.
+
+
+## sed 2/2
+
+A very simple `sed` program:
+```
+sed 's/foo/bar/'
+```
+\
+This program replaces all instance of 'foo' with 'bar' on every line.
+
+\
+Beware of `sed` commands as they can modify a file. 
+
+
+## grep
+
+`grep` will print all lines containing or not containing a string of characters. The following will stream out all input lines containing 'foo':
+\
+```
+grep foo
+```
+\
+
+The following will stream out all input lines not containing 'foo':
+\
+```
+grep -v foo
+```
+
+
+## Exercise 6 {.smaller}
 
 ::: {.panel-tabset}
 
 ### Problem
 
-`shell-training/data/` contains 300 data files, each of which should contain 100 values. One of these files is missing some data though. . .
+`shell-training/data/` contains 300 data files, each of which should contain 100 values. One or more of these files are missing some data though...
 
 \
-Use a series of commands connected by pipes to identify the file with missing data. Hints:
+Use a series of commands connected by pipes to identify which files have missing data. Hints:
 
 - `wc -w` will tell you the number of values in a file
 - `sort -n` will sort numerically
 
 ### Solution
 
+A solution using `grep`:
 ```{.bash code-line-numbers=false}
-[user@host data]$ wc -w values* | sort -n | head -1
-99 values 44
+[user@host data]$ wc -w values* | sort -n | grep -v '100 values'
+    99 values 44
+ 29899 total
 ```
 
-- `wc -w` tells you how many values are in each file
-- `sort` this output to put the files in order of the number of values
-- use `head` to select the first line, which contains the name of the file with the smallest number of values
+\
+Alternatively `awk` could be used:
+```{.bash code-line-numbers=false}
+[user@host data]$ wc -w values* | sort -n | awk '$1 != 100 {print $0}'
+    99 values 44
+ 29899 total
+```
 
 :::
 
 
-## Exercise 5 {.smaller}
+## For loops
+
+Example for loop to print the numbers 1, 3 and 5:
+
+```{.bash code-line-numbers=false}
+[user@host ~]$ for ((i=1;i<6;i=i+2)) do echo $i; done
+1
+3
+5
+```
+\
+`echo` is a command that prints to the standard output.
+The variable `i` is initialised with a value of `1`; the loop continues until the condition `i` is less than `6`; each iteration of the loop `i` is incremented by `2`.
+
+
+## Sequences
+
+Example sequence to print the numbers 1, 3 and 5:
+
+```{.bash code-line-numbers=false}
+[user@host ~]$ for i in {1..5..2}; do echo $i; done
```

Comment:
Should we have introduced a simple sequence {1..5} before introducing a complex one {1..5..2} ?